### ISSUE TITLE
observation_and_action_constraint_splitter doesn't support packed observations in a TFEnvironment per its docstring

### ISSUE THREAD

**Issue by wes-turner** [tensorflow/agents#397]
Documentation for `observation_and_action_constraint_splitter` is replicated throughout the `tf-agents` codebase ([RandomTFPolicy](https://www.tensorflow.org/agents/api_docs/python/tf_agents/policies/random_tf_policy/RandomTFPolicy?hl=he) for example) saying:

> An example observation_and_action_constraint_splitter could be as simple as:
> ```python
>   def observation_and_action_constraint_splitter(observation):
>     return observation['network_input'], observation['constraint']
> ```

I think this makes sense for Python environments, but it doesn't work with TF environments (like those necessarily used in a `RandomTFPolicy`).  In a TF environment, `observation` is a `tf.Tensor` and can't be dereferenced like a `dict` can.

It's (quite) possible I've misunderstood the intended use of splitters altogether, so I'll build up an example, here.  Let's take for that example a Python environment `py_env` and the wrapped, TF version `tf_env = tf_py_environment.TFPyEnvironment(py_env)` each for which the `time_step_spec` defines an `observation` with a 1x2 array of floats and the `action_spec` a scalar int between 0 and 9 inclusive.

In line with the docstring's suggested splitter, an observation as returned by `py_env.reset()` looks like:
```python
TimeStep(
  step_type=array(0, dtype=int32),
  reward=array(0., dtype=float32),
  discount=array(1., dtype=float32),
  observation={
    'network_input': array([0.90490425, 0.21128751], dtype=float32),
    'constraint': [True, True, True, True, True, True, True, True, True, True]})
```

`py_env` was created with `suite_gym.load` and a `wrappers.PyEnvironmentBaseWrapper` that transforms the 1x2 observation array returned by the gym environment's reset function into the datatype above.

`tf_env.reset()`, by contrast, happily and silently returns the following `TimeStep`:
```python
TimeStep(
  step_type=<tf.Tensor: shape=(1,), dtype=int32, numpy=array([0], dtype=int32)>, 
  reward=<tf.Tensor: shape=(1,), dtype=float32, numpy=array([0.], dtype=float32)>, 
  discount=<tf.Tensor: shape=(1,), dtype=float32, numpy=array([1.], dtype=float32)>, 
  observation=<tf.Tensor: shape=(1,), dtype=bool, numpy=array([ True])>)
```
FWIW, the observation is reshaped by a large stack of function calls, started by `batch_nested_array` as called [here](https://github.com/tensorflow/agents/blob/720a7bc9e7519d1ca7749fa71189e9c3aa1a5483/tf_agents/environments/batched_py_environment.py#L146), which transforms the observation to:
```python
  observation={
    'network_input': array([[0.5526981 , 0.20855443]], dtype=float32),
    'constraint': [array([ True]), array([ True]), array([ True]), array([ True]), array([ True]), array([ True]), array([ True]), array([ True]), array([ True]), array([ True])]}
```
Is there alternate guidance for a canonical observation data structure (and splitter) for use in both `PyEnvironment`s and `TFEnvironment`s?  `observation_and_action_constraint_splitter` appears to be an infrequently-used feature, and what discussion I've seen about it tends to define a splitter that takes an `observation` as its only input.  Great when the action constraint is easy to compute directly from an observation, but not so broadly applicable as the docstring's approach to wrap an observation in a dict.

Or instead, is there a better approach altogether for collecting action constraints from an underlying gym environment?

**Comment by ebrevdo** [comment]
@ormandi can you ptal?

**Comment by ormandi** [comment]
Ah, haven't noticed this earlier. Taking a look tomorrow morning.

**Comment by ormandi** [comment]
Hi, Wes.

Thank you for opening an issue, and terribly sorry for the long turnaround on this.

TL;DR: Actually, both `py_env` and `tf_env` should support arbitrary nested observation space spec, and the conversion from `py_env` to `tf_env` definitely should not change the *structure* of the input spec which seems to be failing in your case. Once this is fixed, I think the splitter will work just like in the case of the `py_env`

Here is a quick repro of the expected behavior using the example `CardGameEnv` from [https://github.com/tensorflow/agents/blob/master/docs/tutorials/2_environments_tutorial.ipynb](https://github.com/tensorflow/agents/blob/master/docs/tutorials/2_environments_tutorial.ipynb) after changing the observation spec to be a `dict` with two entries (`observation`, and `mask`; please see the whole environment at the end of the comment). Given this environment the, the python env should look like:
```python
> py_env = CardGameEnvWithMask()
> utils.validate_py_environment(py_env, episodes=5)
> print(py_env.observation_spec())
{
  'observation': 
      BoundedArraySpec(shape=(1,), dtype=dtype('int32'), name='observation', minimum=0, maximum=2147483647),
  'mask': ArraySpec(shape=(1,), dtype=dtype('bool'), name='mask')
}
> py_ts = py_env.reset()
> print(py_ts)
TimeStep(
  step_type=array(0, dtype=int32),
  reward=array(0., dtype=float32), discount=array(1., dtype=float32),
  observation={
      'observation': array([0], dtype=int32),
      'mask': array([ True])
  })
```
while the converted TF env should be:
```python
> tf_env = tf_py_environment.TFPyEnvironment(py_env)
> print(tf_env.observation_spec())
{
  'observation':
      BoundedTensorSpec(shape=(1,), dtype=tf.int32, name='observation', minimum=array(0, dtype=int32), maximum=array(2147483647, dtype=int32)),
  'mask': TensorSpec(shape=(1,), dtype=tf.bool, name='mask')
}
> tf_ts = tf_env.reset()
> print(tf_ts)
TimeStep(
  step_type=<tf.Tensor: shape=(1,), dtype=int32, numpy=array([0], dtype=int32)>,
  reward=<tf.Tensor: shape=(1,), dtype=float32, numpy=array([0.], dtype=float32)>,
  discount=<tf.Tensor: shape=(1,), dtype=float32, numpy=array([1.], dtype=float32)>,
  observation={
      'observation': <tf.Tensor: shape=(1, 1), dtype=int32, numpy=array([[0]], dtype=int32)>,
      'mask': <tf.Tensor: shape=(1, 1), dtype=bool, numpy=array([[ True]])>
})
```
As you can see in both cases the observation is a `dict` (so the structure not, but the elements of the nest did change), so applying the same splitter should work just fine.

My suggestion would be to try to take a look into the environment, and make sure that it is valid (you may want to check out `utils.validate_py_environment`), and the conversion happens as expected.

Regarding the mask splitter functionality: There is a `MaskSplitterNetwork` implementation (https://github.com/tensorflow/agents/blob/master/tf_agents/networks/mask_splitter_network.py) which pushes the splitting logic into a network wrapper, so arbitrary network can be used with splitter. However, this class relies on a very similar splitter function, so the environment has to be fixed first.

Hope this helps.

Finally, the slightly extended `CardGameEnv` containing a mask:
```python
# The original env (without the masking part) is from TF-Agents Environments
# Colab Notebook available at:
# https://github.com/tensorflow/agents/blob/master/docs/tutorials/2_environments_tutorial.ipynb
class CardGameEnvWithMask(py_environment.PyEnvironment):

  def __init__(self):
    self._action_spec = array_spec.BoundedArraySpec(
        shape=(), dtype=np.int32, minimum=0, maximum=1, name='action')
    self._observation_spec = {
        'observation': array_spec.BoundedArraySpec(
            shape=(1,), dtype=np.int32, minimum=0, name='observation'),
        'mask': array_spec.ArraySpec(
            shape=(1,), dtype=np.bool, name='mask')}
    self._state = 0
    self._episode_ended = False

  def action_spec(self):
    return self._action_spec

  def observation_spec(self):
    return self._observation_spec

  def _obs(self, obs, mask=True):
    return {'observation': obs, 'mask': np.array([mask], dtype=np.bool)}

  def _reset(self):
    self._state = 0
    self._episode_ended = False
    return ts.restart(self._obs(np.array([self._state], dtype=np.int32)))

  def _step(self, action):

    if self._episode_ended:
      # The last action ended the episode. Ignore the current action and start
      # a new episode.
      return self.reset()

    # Make sure episodes don't go on forever.
    if action == 1:
      self._episode_ended = True
    elif action == 0:
      new_card = np.random.randint(1, 11)
      self._state += new_card
    else:
      raise ValueError('`action` should be 0 or 1.')
    
    if self._episode_ended or self._state >= 21:
      reward = self._state - 21 if self._state <= 21 else -21
      return ts.termination(
          self._obs(np.array([self._state], dtype=np.int32), False), reward)
    else:
      return ts.transition(
          self._obs(np.array([self._state], dtype=np.int32), True),
          reward=0.0, discount=1.0)
```

**Comment by tagomatech** [comment]
Apologies in advance for reviving this closed issue with the maybe irrelevant comment that follows.

I came across @wes-turner's issue above while writing myself a python environment with mixed-type observations, Using @ormandi's code as a template, `self._observation_spec` would be a 2-column array, hence of `shape=(2,)` (if I am being correct), 1 element would be a `np.float64 `and the other one a `np.int32`.

At a glance, I though the solution provided by @ormandi would do the trick. But, I am puzzled because `utils.validate_py_environment` fails to validate the `CardGameEnvWithMask` environment proposed. More specificaly, I am returned:

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-23-4fbfaa7bef9e> in <module>
     53 
     54 env = TradingEnv()
---> 55 utils.validate_py_environment(environment=env, episodes=10)

c:\program files\python38\lib\site-packages\tf_agents\environments\utils.py in validate_py_environment(environment, episodes)
     59 
     60   episode_count = 0
---> 61   time_step = environment.reset()
     62 
     63   while episode_count < episodes:

c:\program files\python38\lib\site-packages\tf_agents\environments\py_environment.py in reset(self)
    168           corresponding to `observation_spec()`.
    169     """
--> 170     self._current_time_step = self._reset()
    171     return self._current_time_step
    172 

<ipython-input-22-cb6df2f0b0a7> in _reset(self)
     50     def _reset(self):
     51         self_state = 0
---> 52         return ts.restart(np.array([self._state], dtype=np.int32))
     53 
     54     def _step(self, action):

TypeError: int() argument must be a string, a bytes-like object or a number, 'not set'
```
And actually, looking into [tf_agents/environments/utils.py](https://github.com/tensorflow/agents/blob/master/tf_agents/environments/utils.py), it makes sense.

What would be your recommendations, then? And what is the right way to define `self._observation_spec` with mixed type in `tf-agents`?

Hopefully the above is sensical from both `RL` and `tf-agents`' standpoints.

Many thanks.
tag

**Comment by ormandi** [comment]
Hello, thank you so much for your comment.

Would it be possible to define what exactly mean by "python environment with mixed-type observations"? Or, even better, can you share (a relevant piece of) your code to make it easier identify the culprit?

In general, an observation can be an arbitrary nested structure. This structure is defined by the person who implements the environment in order to help the readability/understanding of the observation space. In this context, it seems `shape=(2,)` is in the leaf of the nested structure meaning that the particular leaf is a 1D array-like object with size 2. Within this, I can't see how the type could be varied. Do I understand this setup correctly? If so, is it possible to split this up into e.g. a tuple of two objects with `shap=(1,)` for each but with different types?

Sorry for my vague answer.

**Comment by tagomatech** [comment]
@ormandi Thank you for kindly adding new insights.
With "mixed-type observations", I meant the agent e.g. is a car driver who observes whether it is raining (first observation is boolean) and car speed (second observation is a float). So we would have "mixed-type observations.
What you wrote down above suggests that I go for a nested structure that would (in my car driver toy example) contains a boolean and a float (?).

**Comment by ormandi** [comment]
Thanks again for the feedback. Yes, I think something like a tuple of two scalars one being `bool` while the other `float` will do the job.
